modules/seqexec/server/src/main/scala/seqexec/server/SeqTranslate.scala

```diff
@@ -487,7 +488,7 @@ object SeqTranslate {
     def toAction(kind: ActionType): Action[F] = fromF[F](kind, x.attempt.map(_.toResult))
   }
 
-  implicit class ConfigResultToAction[F[_]: Functor: ApplicativeError[?[_], Throwable]](val x: F[ConfigResult[F]]) {
```

Comment:
This is something I don't recall having seen before. What is a `?[_]`?

Reply:
It means (i think) you want an `ApplicativeError` for some `F[_]`. Maybe @tpolecat can explain it ibetter

Reply:
Yeah the syntax is a little wacky.

A context bound `[X : Y]` desugars to `(implicit ev: Y[X])` where `ev` is a unique inaccessible name the compiler generates. So in order for this to be well-kinded `Y` must be a type constructor (i.e., a type-level function) with a single parameter of the same kind ("shape") as `X`.

A first-order example would be

- `[A : Monoid]` which is how you normally write it, or
- `[A : Monoid[?]]` which shows the hole explicitly.

These are equivalent. A higher-order example would be

- `[F[_] : Functor]` which is how you normally write it, or
- `[F[_] : Functor[?[_]]]` which shows the hole explicitly.

Note that the hole `?[_]` has the same shape as `F[_]`. They have to match, otherwise it's ill-kinded and you'll get a compile error.

Ok, so what we're dealing with here is `ApplicativeError[F[_], E]` which has *two* parameters, but a context bound requires a type constructor with exactly one. So you have to partially apply it, and to do this you need the second form, where we show the hole explicitly for the type variable we leaving free, and specify `Throwable` for the other one.

- `[F[_]: ApplicativeError[?[_], Throwable]]`

Which in the end desugars to `(implicit ev: ApplicativeError[F, Throwable])`

Make sense? When in doubt you can always write out the implicit parameter list. This is just shorthand.

Reply:
@tpolecat Thanks for the explanation. This works only because we have the `kind-projector` plugin right? What would the story be in dotty?

Reply:
Something very close to kind-projector is built into dotty but I don't know exactly how you would write this.